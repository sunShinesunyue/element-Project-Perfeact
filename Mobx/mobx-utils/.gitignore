node_modules
*.log
lib
coverage
mobx-utils.module.js
mobx-utils.umd.js
.idea
/.test-ts/
*.iml
*.ipr
*.iws
