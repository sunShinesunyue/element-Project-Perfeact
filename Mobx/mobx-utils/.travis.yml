language: node_js
node_js:
  - 8
cache:
  directories:
    - node_modules
install:
  - yarn install
script:
  - yarn run coverage
after_success:
  - cat ./coverage/lcov.info|./node_modules/coveralls/bin/coveralls.js
