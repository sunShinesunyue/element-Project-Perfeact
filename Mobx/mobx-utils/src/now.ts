import { _isComputingDerivation } from "mobx"
import { fromResource, IResource } from "./from-resource"

const tickers: {
    [interval: string]: IResource<number>
} = {}

/**
 * Returns the current date time as epoch number.
 * The date time is read from an observable which is updated automatically after the given interval.
 * So basically it treats time as an observable.
 *
 * The function takes an interval as parameter, which indicates how often `now()` will return a new value.
 * If no interval is given, it will update each second. If "frame" is specified, it will update each time a
 * `requestAnimationFrame` is available.
 *
 * Multiple clocks with the same interval will automatically be synchronized.
 *
 * Countdown example: https://jsfiddle.net/mweststrate/na0qdmkw/
 *
 * @example
 *
 * const start = Date.now()
 *
 * autorun(() => {
 *   console.log("Seconds elapsed: ", (mobxUtils.now() - start) / 1000)
 * })
 *
 *
 * @export
 * @param {(number | "frame")} [interval=1000] interval in milliseconds about how often the interval should update
 * @returns
 */
export function now(interval: number | "frame" = 1000) {
    if (!_isComputingDerivation()) {
        // See #40
        return Date.now()
    }
    if (!tickers[interval]) {
        if (typeof interval === "number") tickers[interval] = createIntervalTicker(interval)
        else tickers[interval] = createAnimationFrameTicker()
    }
    return tickers[interval].current()
}

function createIntervalTicker(interval: number): IResource<number> {
    let subscriptionHandle: any
    return fromResource<number>(
        sink => {
            subscriptionHandle = setInterval(() => sink(Date.now()), interval)
        },
        () => {
            clearInterval(subscriptionHandle)
        },
        Date.now()
    )
}

function createAnimationFrameTicker(): IResource<number> {
    let subscriptionHandle: number
    const frameBasedTicker = fromResource<number>(
        sink => {
            function scheduleTick() {
                window.requestAnimationFrame(() => {
                    sink(Date.now())
                    if (frameBasedTicker.isAlive()) scheduleTick()
                })
            }
            scheduleTick()
        },
        () => {},
        Date.now()
    )
    return frameBasedTicker
}
