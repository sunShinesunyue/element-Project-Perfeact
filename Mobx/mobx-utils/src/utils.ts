export type IDisposer = () => void

export const NOOP = () => {}

export const IDENTITY = (_: any) => _

export function invariant(cond: boolean, message = "Illegal state") {
    if (!cond) throw new Error("[mobx-utils] " + message)
}

const deprecatedMessages: string[] = []
export function deprecated(msg: string) {
    if (deprecatedMessages.indexOf(msg) !== -1) return
    deprecatedMessages.push(msg)
    console.error("[mobx-utils] Deprecated: " + msg)
}

export function addHiddenProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: true,
        configurable: true,
        value
    })
}
