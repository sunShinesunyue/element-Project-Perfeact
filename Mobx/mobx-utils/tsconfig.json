{
    "version": "1.8.10",
    "compilerOptions": {
        "target": "es5",
        "module": "commonjs",
        "moduleResolution": "node",
        "experimentalDecorators": true,
        "noEmit": true,
        "rootDir": ".",
        "lib": [
            "dom", "es5", "scripthost", "es2015.promise", "es2015.generator", "es2015.iterable"
        ]
    },
    "include": [
        "**/*.ts"
    ],
    "exclude": [
        "../node_modules"
    ]
}
