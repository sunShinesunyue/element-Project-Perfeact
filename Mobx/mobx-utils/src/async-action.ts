import { flow } from "mobx"
import { deprecated } from "./utils"

// method decorator:
export function asyncAction(
    target: Object,
    propertyKey: string,
    descriptor: PropertyDescriptor
): PropertyDescriptor

// non-decorator forms
export function asyncAction<R>(generator: () => IterableIterator<any>): () => Promise<R>
export function asyncAction<A1>(
    generator: (a1: A1) => IterableIterator<any>
): (a1: A1) => Promise<any> // Ideally we want to have R instead of Any, but cannot specify R without specifying A1 etc... 'any' as result is better then not specifying request args
export function asyncAction<A1, A2, A3, A4, A5, A6, A7, A8>(
    generator: (
        a1: A1,
        a2: A2,
        a3: A3,
        a4: A4,
        a5: A5,
        a6: A6,
        a7: A7,
        a8: A8
    ) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => Promise<any>
export function asyncAction<A1, A2, A3, A4, A5, A6, A7>(
    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => Promise<any>
export function asyncAction<A1, A2, A3, A4, A5, A6>(
    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => Promise<any>
export function asyncAction<A1, A2, A3, A4, A5>(
    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => Promise<any>
export function asyncAction<A1, A2, A3, A4>(
    generator: (a1: A1, a2: A2, a3: A3, a4: A4) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4) => Promise<any>
export function asyncAction<A1, A2, A3>(
    generator: (a1: A1, a2: A2, a3: A3) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3) => Promise<any>
export function asyncAction<A1, A2>(
    generator: (a1: A1, a2: A2) => IterableIterator<any>
): (a1: A1, a2: A2) => Promise<any>
export function asyncAction<A1>(
    generator: (a1: A1) => IterableIterator<any>
): (a1: A1) => Promise<any>
// ... with name
export function asyncAction<R>(
    name: string,
    generator: () => IterableIterator<any>
): () => Promise<R>
export function asyncAction<A1>(
    name: string,
    generator: (a1: A1) => IterableIterator<any>
): (a1: A1) => Promise<any> // Ideally we want to have R instead of Any, but cannot specify R without specifying A1 etc... 'any' as result is better then not specifying request args
export function asyncAction<A1, A2, A3, A4, A5, A6, A7, A8>(
    name: string,
    generator: (
        a1: A1,
        a2: A2,
        a3: A3,
        a4: A4,
        a5: A5,
        a6: A6,
        a7: A7,
        a8: A8
    ) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => Promise<any>
export function asyncAction<A1, A2, A3, A4, A5, A6, A7>(
    name: string,
    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => Promise<any>
export function asyncAction<A1, A2, A3, A4, A5, A6>(
    name: string,
    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => Promise<any>
export function asyncAction<A1, A2, A3, A4, A5>(
    name: string,
    generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => Promise<any>
export function asyncAction<A1, A2, A3, A4>(
    name: string,
    generator: (a1: A1, a2: A2, a3: A3, a4: A4) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3, a4: A4) => Promise<any>
export function asyncAction<A1, A2, A3>(
    name: string,
    generator: (a1: A1, a2: A2, a3: A3) => IterableIterator<any>
): (a1: A1, a2: A2, a3: A3) => Promise<any>
export function asyncAction<A1, A2>(
    name: string,
    generator: (a1: A1, a2: A2) => IterableIterator<any>
): (a1: A1, a2: A2) => Promise<any>
export function asyncAction<A1>(
    name: string,
    generator: (a1: A1) => IterableIterator<any>
): (a1: A1) => Promise<any>

/**
 * _deprecated_ this functionality can now be found as `flow` in the mobx package. However, `flow` is not applicable as decorator, where `asyncAction` still is.
 *
 *
 *
 * `asyncAction` takes a generator function and automatically wraps all parts of the process in actions. See the examples below.
 * `asyncAction` can be used both as decorator or to wrap functions.
 *
 * - It is important that `asyncAction should always be used with a generator function (recognizable as `function*` or `*name` syntax)
 * - Each yield statement should return a Promise. The generator function will continue as soon as the promise settles, with the settled value
 * - When the generator function finishes, you can return a normal value. The `asyncAction` wrapped function will always produce a promise delivering that value.
 *
 * When using the mobx devTools, an asyncAction will emit `action` events with names like:
 * * `"fetchUsers - runid: 6 - init"`
 * * `"fetchUsers - runid: 6 - yield 0"`
 * * `"fetchUsers - runid: 6 - yield 1"`
 *
 * The `runId` represents the generator instance. In other words, if `fetchUsers` is invoked multiple times concurrently, the events with the same `runid` belong toghether.
 * The `yield` number indicates the progress of the generator. `init` indicates spawning (it won't do anything, but you can find the original arguments of the `asyncAction` here).
 * `yield 0` ... `yield n` indicates the code block that is now being executed. `yield 0` is before the first `yield`, `yield 1` after the first one etc. Note that yield numbers are not determined lexically but by the runtime flow.
 *
 * `asyncActions` requires `Promise` and `generators` to be available on the target environment. Polyfill `Promise` if needed. Both TypeScript and Babel can compile generator functions down to ES5.
 *
 *  N.B. due to a [babel limitation](https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy/issues/26), in Babel generatos cannot be combined with decorators. See also [#70](https://github.com/mobxjs/mobx-utils/issues/70)
 *
 *
 * @example
 * import {asyncAction} from "mobx-utils"
 *
 * let users = []
 *
 * const fetchUsers = asyncAction("fetchUsers", function* (url) {
 *   const start = Date.now()
 *   const data = yield window.fetch(url)
 *   users = yield data.json()
 *   return start - Date.now()
 * })
 *
 * fetchUsers("http://users.com").then(time => {
 *   console.dir("Got users", users, "in ", time, "ms")
 * })
 *
 * @example
 * import {asyncAction} from "mobx-utils"
 *
 * mobx.configure({ enforceActions: "observed" }) // don't allow state modifications outside actions
 *
 * class Store {
 * 	\@observable githubProjects = []
 * 	\@state = "pending" // "pending" / "done" / "error"
 *
 * 	\@asyncAction
 * 	*fetchProjects() { // <- note the star, this a generator function!
 * 		this.githubProjects = []
 * 		this.state = "pending"
 * 		try {
 * 			const projects = yield fetchGithubProjectsSomehow() // yield instead of await
 * 			const filteredProjects = somePreprocessing(projects)
 * 			// the asynchronous blocks will automatically be wrapped actions
 * 			this.state = "done"
 * 			this.githubProjects = filteredProjects
 * 		} catch (error) {
 * 			this.state = "error"
 * 		}
 * 	}
 * }
 *
 * @export
 * @returns {Promise}
 */

export function asyncAction(arg1: any, arg2?: any): any {
    // decorator
    if (typeof arguments[1] === "string") {
        const name = arguments[1]
        const descriptor: PropertyDescriptor = arguments[2]
        if (descriptor && descriptor.value) {
            return Object.assign({}, descriptor, {
                value: flow(descriptor.value)
            })
        } else {
            return Object.assign({}, descriptor, {
                set(v: any) {
                    Object.defineProperty(this, name, {
                        ...descriptor,
                        value: flow(v)
                    })
                }
            })
        }
    }

    // direct invocation
    const generator = typeof arg1 === "string" ? arg2 : arg1
    const name = typeof arg1 === "string" ? arg1 : generator.name || "<unnamed async action>"
    deprecated("asyncAction is deprecated. use mobx.flow instead")
    return flow(generator) // name get's dropped..
}
