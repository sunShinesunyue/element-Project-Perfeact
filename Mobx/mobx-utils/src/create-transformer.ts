import { computed, onBecomeUnobserved, IComputedValue } from "mobx"
import { invariant, addHiddenProp } from "./utils"

export type ITransformer<A, B> = (object: A) => B

let memoizationId = 0

/**
 * Creates a function that maps an object to a view.
 * The mapping is memoized.
 *
 * See: https://mobx.js.org/refguide/create-transformer.html
 */
export function createTransformer<A, B>(
    transformer: ITransformer<A, B>,
    onCleanup?: (resultObject: B | undefined, sourceObject?: A) => void
): ITransformer<A, B> {
    invariant(
        typeof transformer === "function" && transformer.length < 2,
        "createTransformer expects a function that accepts one argument"
    )

    // Memoizes: object id -> reactive view that applies transformer to the object
    let views: { [id: number]: IComputedValue<B> } = {}

    function createView(sourceIdentifier: number, sourceObject: A) {
        let latestValue: B
        const expr = computed(
            () => {
                return (latestValue = transformer(sourceObject))
            },
            {
                name: `Transformer-${(<any>transformer).name}-${sourceIdentifier}`
            }
        )
        const disposer = onBecomeUnobserved(expr, () => {
            delete views[sourceIdentifier]
            disposer()
            if (onCleanup) onCleanup(latestValue, sourceObject)
        })
        return expr
    }

    return (object: A) => {
        const identifier = getMemoizationId(object)
        let reactiveView = views[identifier]
        if (reactiveView) return reactiveView.get()
        // Not in cache; create a reactive view
        reactiveView = views[identifier] = createView(identifier, object)
        return reactiveView.get()
    }
}

function getMemoizationId(object: any) {
    const objectType = typeof object
    if (objectType === "string") return `string:${object}`
    if (objectType === "number") return `number:${object}`
    if (object === null || (objectType !== "object" && objectType !== "function"))
        throw new Error(
            `[mobx-utils] transform expected an object, function, string or number, got: ${String(object)}`
        )
    let tid = object.$transformId
    if (tid === undefined) {
        tid = `memoizationId:${++memoizationId}`
        addHiddenProp(object, "$transformId", tid)
    }
    return tid
}
