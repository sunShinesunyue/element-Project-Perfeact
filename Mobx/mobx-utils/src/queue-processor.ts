import { isAction, autorun, action, isObservableArray, runInAction } from "mobx"
import { IDisposer } from "./utils"

/**
 * `queueProcessor` takes an observable array, observes it and calls `processor`
 * once for each item added to the observable array, optionally deboucing the action
 *
 * @example
 * const pendingNotifications = observable([])
 * const stop = queueProcessor(pendingNotifications, msg => {
 *   // show Desktop notification
 *   new Notification(msg);
 * })
 *
 * // usage:
 * pendingNotifications.push("test!")
 *
 * @param {T[]} observableArray observable array instance to track
 * @param {(item: T) => void} processor action to call per item
 * @param {number} [debounce=0] optional debounce time in ms. With debounce 0 the processor will run synchronously
 * @returns {IDisposer} stops the processor
 */
export function queueProcessor<T>(
    observableArray: T[],
    processor: (item: T) => void,
    debounce = 0
): IDisposer {
    if (!isObservableArray(observableArray))
        throw new Error("Expected observable array as first argument")
    if (!isAction(processor)) processor = action("queueProcessor", processor)

    const runner = () => {
        // construct a final set
        const items = observableArray.slice(0)
        // clear the queue for next iteration
        runInAction(() => observableArray.splice(0))
        // fire processor
        items.forEach(processor)
    }
    if (debounce > 0) return autorun(runner, { delay: debounce })
    else return autorun(runner)
}
