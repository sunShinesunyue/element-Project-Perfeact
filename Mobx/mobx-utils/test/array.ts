import * as utils from "../src/mobx-utils"
import { observable } from "mobx"
import { moveItem } from "../src/mobx-utils"

test("it should move the item as expected", () => {
    const source = observable<number>([1, 2, 3])
    expect(moveItem(source, 0, 1)).toBe(source)
    expect(source[0]).toBe(2)
    expect(source[1]).toBe(1)
    expect(source[2]).toBe(3)

    moveItem(source, 1, 0)

    expect(source[0]).toBe(1)
    expect(source[1]).toBe(2)
    expect(source[2]).toBe(3)
})
test("it throws when index is out of bounds", () => {
    const source = observable<number>([1, 2, 3])
    expect(moveItem(source, 0, 0)).toBeUndefined()
    expect(moveItem(source, 2, 2)).toBeUndefined()
})

test("it throws when index is out of bounds", () => {
    const source = observable<number>([1, 2, 3])
    expect(() => moveItem(source, 0, -1)).toThrowErrorMatchingSnapshot()
    expect(() => moveItem(source, 0, 3)).toThrowErrorMatchingSnapshot()
})
