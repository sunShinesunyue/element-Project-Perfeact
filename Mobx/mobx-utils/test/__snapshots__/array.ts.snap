// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it throws when index is out of bounds 1`] = `"[mobx.array] Index out of bounds: -1 is negative"`;

exports[`it throws when index is out of bounds 2`] = `"[mobx.array] Index out of bounds: 3 is not smaller than 3"`;
